From 579c417598ac24c60ef12a762e67a07ab1046234 Mon Sep 17 00:00:00 2001
From: saleemuddin shaik <quic_saleemud@quicinc.com>
Date: Tue, 7 Mar 2023 16:47:53 +0530
Subject: [PATCH] hostapd: add vendor_oui and UNII band into sta_info

multiple calls of WPA_GET_BE24 and right shifting avoided.
Introduced condition to extract vendor oui by looping through the ies.

Signed-off-by: saleemuddin shaik <quic_saleemud@quicinc.com>
---
 src/ap/ctrl_iface_ap.c | 6 ++++++
 src/ap/ieee802_11.c    | 9 +++++++++
 src/ap/sta_info.h      | 3 +++
 3 files changed, 18 insertions(+)

--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -327,6 +327,12 @@ static int hostapd_ctrl_iface_sta_mib(st
 		len += ret;
 	}
 
+	ret = os_snprintf(buf + len, buflen - len, "vendor_oui=%02x:%02x:%02x\n",
+			  sta->vendor_oui[0], sta->vendor_oui[1], sta->vendor_oui[2]);
+	if (os_snprintf_error(buflen - len, ret))
+		return len;
+	len += ret;
+
 	res = ieee802_11_get_mib_sta(hapd, sta, buf + len, buflen - len);
 	if (res >= 0)
 		len += res;
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -3973,6 +3973,15 @@ static int __check_assoc_ies(struct host
 	struct hostapd_data *assoc_hapd;
 	struct sta_info *assoc_sta = NULL;
 
+	const struct element *elem;
+
+	for_each_element(elem, ies, ies_len) {
+		memcpy(sta->vendor_oui, elem->data, 3);
+		if(elem->id == WLAN_EID_VENDOR_SPECIFIC &&
+		   !WPA_GET_BE24(sta->vendor_oui))
+			break;
+	}
+
 	resp = check_ssid(hapd, sta, elems->ssid, elems->ssid_len);
 	if (resp != WLAN_STATUS_SUCCESS)
 		return resp;
--- a/src/ap/sta_info.h
+++ b/src/ap/sta_info.h
@@ -317,6 +317,9 @@ struct sta_info {
 	struct pasn_data *pasn;
 #endif /* CONFIG_PASN */
 
+	/* Vendor Specific OUI from associated STA */
+	u8 vendor_oui[3];
+
 #ifdef CONFIG_IEEE80211BE
 	struct mld_info mld_info;
 	u8 mld_assoc_link_id;
